src/viam/trajex/totg/trajectory.cpp

```diff
@@ -439,40 +440,123 @@ enum class integration_event : std::uint8_t {
 }
 
 // Searches for a velocity switching point where escape from velocity curve becomes possible.
-// When trapped on the velocity limit curve, we search forward to find where the trajectory
-// can drop below the curve (equation 40). Uses coarse search followed by bisection refinement.
-// See Kunz & Stilman Section VII-B, equations 40-42.
+// Implements both continuous (equation 40) and discontinuous (equations 41-42) cases from Section VII-B.
 // Returns the switching point on velocity curve. If no escape point found, returns end of path.
 // Takes path::cursor by value (cheap copy) to seed forward search from current position.
 [[gnu::pure]] trajectory::phase_point find_velocity_switching_point(path::cursor cursor, const trajectory::options& opt) {
     const arc_length path_length = cursor.path().length();
-    const double step_size = opt.delta.count();
 
-    // Phase 1: Coarse forward search for escape condition (equation 40).
+    std::optional<trajectory::phase_point> discontinuous_switching_point;
+    std::optional<trajectory::phase_point> continuous_switching_point;
+
+    // Phase 1: Check for discontinuous switching points at segment boundaries (equations 41-42).
+    // Section VII-B case 2: f''_i(s) being discontinuous is a necessary condition for
+    // s_ddot_min(s, s_dot_max_vel(s)) being discontinuous.
+    // We need to check ALL boundaries to find the earliest one, not return immediately.
```

Comment:
I'm not going to suggest we do it now, but this feels to me like we really shouldn't have separate `find_velocity_switching_point` and `find_acceleration_switching_point` functions. We should just have a `find_switching_point`. Both of them are doing checks at segment boundaries. Both of them are looking for multiple kinds of switching points. Both of them have early exit conditions. And, in the end, we need to run both algorithms and select one winner.

It'd be a complicated function to be sure, but it would probably be faster.

Reply:
https://viam.atlassian.net/browse/RSDK-12819